include/gabia/secure_socket.hpp

```diff
@@ -0,0 +1,302 @@
+
+    template <typename ConstBufferSequence>
+    auto write_some(const ConstBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = write_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence,
+                          beast::error_code& error) {
+        error = {};
+        if (decrypted_buffer.size() > 0) {
+            // We have already decrypted data ready to be retrieved from the
+            // buffer
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            return consumed;
+        }
+
+        boost::endian::little_int16_t payload_length = 0;
+        size_t transferred = 0;
+        while (transferred < sizeof(payload_length) + payload_length +
+                                 crypto::aead_auth_tag_size) {
+            auto mb = raw_buffer.prepare(max_payload_length);
+            transferred += next_layer().read_some(mb, error);
+            if (error) {
+                return 0;
+            }
+            raw_buffer.commit(transferred);
+            extract_pdu_length(payload_length);
+        }
+
+        auto success = decrypt_pdu(payload_length);
+        if (!success) {
+            // Forged/spoofed message
+            // TODO: use a more appropriate error code
+            error = asio::error::no_permission;
+            decrypted_buffer.consume(payload_length);
+            return 0;
+        }
+
+        auto consumed =
+            boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+        decrypted_buffer.consume(consumed);
+        return consumed;
+    }
+
+    template <typename MutableBufferSequence>
+    std::size_t read_some(const MutableBufferSequence& sequence) {
+        beast::error_code error = {};
+        auto n = read_some(sequence, error);
+        Expects(!error);
+        return n;
+    }
+
+    template <typename MutableBufferSequence, typename Handler>
+    auto async_read_some(MutableBufferSequence sequence, Handler&& handler) {
+        beast::async_completion<Handler, void(beast::error_code, size_t)> init{
+            handler};
+        if (decrypted_buffer.size() > 0) {
+            auto consumed =
+                boost::asio::buffer_copy(sequence, decrypted_buffer.data());
+            decrypted_buffer.consume(consumed);
+            next_layer().get_io_service().dispatch([
+                consumed, handler = std::move(init.completion_handler)
+            ]() mutable {
+                beast::error_code error{};
+                handler(error, consumed);
+            });
+        } else {
+            async_read_some_op<MutableBufferSequence, Handler> operation{
+                *this, sequence, std::move(init.completion_handler)};
+            next_layer().get_io_service().dispatch([operation = std::move(
```

Comment:
Multiple problems. First, the correct type is not `Handler`, it is `beast::handler_type<Handler, void(error_code, std::size_t)>`. Otherwise, your function won't work when used with coroutines or futures.

There is no need to call `dispatch` here, you could simply write:
```
async_read_some_op<
    MutableBufferSequence,
    beast::handler_type<Handler, void(beast::error_code, std::size_t)>>{
        *this,
        sequence,
        std::move(init.completion_handler)}();
```

Note the `()` at the end. To make this compile just give `async_read_some_op::operator()` some defaulted parameters.

By the way I'm not sure its okay to move the completion handler in. Keep in mind this will likely change with the next Boost.Asio update (they have move-only handlers).